Add `Parameters#expect` to safely filter and require params

martinemde opened the pull request:
### Motivation / Background

I've been hunting around trying to fix the problem with the default, recommended way of handling parameters in Rails.

```ruby
user_params = params.require(:user).permit(:name, :age)
```

This is fine until someone using your app starts messing with the parameters and causing 500 errors by passing:

```
/path?user=string
```

This causes a NoMethodError because `permit` is called on `"string"`.

The recommendation is not the best way. Instead, the first statement should be written like this:

```ruby
user_params = params.permit(user: [:name, :age]).require(:user)
```

### Detail

In this PR, I propose a new method on ActionController::Parameters that safely handles params filtering in one method, cuts down on ignorable errors caused by params tampering, and improves the specificity of param types.

```ruby
params = ActionController::Parameters.new(user: { name: "Martin", age: 40 })
permitted = params.expect(user: %i[name age])
permitted.permitted?   # => true
permitted.has_key?(:name) # => true
permitted.has_key?(:age) # => true
```

In order to ensure consistent filtering of hashes vs arrays in params, `expect` introduces a new Array matching syntax using a double array `[[:attr]]`. (this example uses shorthand for all but the first example to make it more readable. The hashes returned are actually permitted ActionController::Parameter objects)

```ruby
# Array is expected with `permit`
ActionController::Parameters.new(comments: { text: "Hello" }).permit(comments: [:text]).require(:comments)
# => #<ActionController::Parameters {"text"=>"Hello"} permitted: true>
ActionController::Parameters.new(comments: [{ text: "Hello" }]).permit(comments: [:text]).require(:comments)
# => [{"text"=>"Hello"}]
Parameters.new(comments: { text: "Hello" }).permit(comments: [[:text]]).require(:comments)
# => param is missing or the value is empty: comments (ActionController::ParameterMissing)
Parameters.new(comments: [{ text: "Hello" }]).permit(comments: [[:text]]).require(:comments)
# => [{"text"=>"Hello"}]

# Array is expected with `expect`
Parameters.new(comments: { text: "Hello" }).expect(comments: [[:text]])
# => param is missing or the value is empty: comments (ActionController::ParameterMissing)
Parameters.new(comments: [{ text: "Hello" }]).expect(comments: [[:text]])
# => [{"text"=>"Hello"}]

# Hash is expected with `permit`
Parameters.new(user: { name: "Martin" }).permit(user: [:name]).require(:user)
# => {"name"=>"Martin"}
Parameters.new(user: [{ name: "Martin" }]).permit(user: [:name]).require(:user)
# => [{"name"=>"Martin"}]

# Hash is expected with `expect`
Parameters.new(user: { name: "Martin" }).expect(user: [:name])
# => {"name"=>"Martin"}
Parameters.new(user: [{ name: "Martin" }]).expect(user: [:name])
# => param is missing or the value is empty: user (ActionController::ParameterMissing)
```

As you can see, `permit` will expect an array when the "double array" syntax is used, but defaults to the old behavior of allowing both arrays and hashes otherwise, maintaining backwards compatibility. `expect` on the other hand is strict in for both arrays and hashes.

This pull request also adds the `expect!` method. `expect!` raises an unhandled error in cases where `expect` would raise a handled error that would return a 400 bad request. The unhandled error can be used in cases where it is important to be alerted about a client using incorrect syntax (such as a private internal API where the client should be fixed).

Adds a 400.html template used to render the 400 bad request response.

### Additional information

Related PRs and issues that address this problem:

* Potentially fixes https://github.com/rails/rails/issues/42953 without changing existing behavior.
* https://github.com/rails/rails/pull/44297 triggered me to dig into this more, where I arrived at this solution.
* Fixes https://github.com/rails/rails/issues/52463
* Fixes https://github.com/rails/rails/issues/42953

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

MatheusRich commented:
@martinemde please add tests and a changelog entry

MatheusRich commented:
I wonder if we should just handle these NoMethodErrors differently. It feels weird to return a 500 when the error is the "the user's fault". Maybe we could return a 4XX somehow?

rafaelfranca commented:
Fixes #51674

rafaelfranca commented:
Let's move on with this one. @dhh already reviewed the API and said it looks promising.

Can you write tests for it and prefer using it to `require.permit` in the guides?

MatheusRich commented:
Fixes #52463

martinemde commented:
I'll update the PR soon

MatheusRich commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -531,6 +531,140 @@ def require(key)
+    # `mandate` is the preferred way to require and permit parameters.
+    # It is similar to calling `permit` and `require` in sequence, but it avoids some
+    # potential pitfalls, such as NoMethodErrors that can be triggered by external
+    # users passing unexpected scalar or non-scalar types.
```
"scalar or non-scalar" doesn't read well IMO. Maybe:

```suggestion
    # users passing unexpected values.
```

```suggestion
    # users passing unexpected param types.
```

martinemde commented:
I'm pushing up work as I finish. Thanks for the feedback. Should have it done soon.

As I've been working on this I thought of an alternate name. What do you think?

```ruby
params.expect(person: %i[name, age])
```

After tying mandate over and over, I'm not totally in love with it, but I don't feel strongly.

dari-us commented:
Seems good so far. I find `expect` sounds nicer, but less strict than `mandate`. When tasked with a name for this from scratch, I'd probably come up with `expect`.

However, seeing as `mandate` internally calls `permit`, it is safe to assume this does not play nicely with `action_on_unpermitted_params`...
According to #26831, it would be fine to have this on a per-controller basis.
Or maybe, remove logging/raising from `permit` and instead implement `permit_only_or_raise`, and `permit_only_or_log` (quite verbose, but the idea is clear)

martinemde commented:
@dari-us maybe `params.expect_only` that would raise.

The existing configuration makes sense. Logging can be useful in dev but would be extra noise in most production applications. This implies to me that logging is an environment config and raising should be controlled by how the methods are called. Please correct me if you think otherwise.

If we added `expect_only` and `permit_only` (for a non-require version) that would allow the use case in #26831. These would be methods that would raise `UnpermittedParameters`. `expect` and `permit` would continue to log or raise depending on the global config, allowing an app that wants both to set globally, and use the `_only` methods selectively. I suggest doing this in a follow-up PR that I can certainly help with.

martinemde commented:
Regarding your comment @MatheusRich, I never got to respond.

> I wonder if we should just handle these NoMethodErrors differently. It feels weird to return a 500 when the error is the "the user's fault". Maybe we could return a 4XX somehow?

This was what motivated me to create this. It's certainly wrong to return 500 when the client caused the problem, but once `require(:key)` returns the value of `params[:key]` we're in plain ruby land and a `NoMethodError` is bound to occur if you call `permit` next. I think we must just remove all uses or mentions of the pattern `require(:key).permit(:attr)` and switch to this new method. I aim to do that here to the extent possible.

`permit(key: [:attr]).require(:key)` is how rubygems.org now processes _all_ params and it has removed an annoying class of 500s that used to get us paged occasionally. I will to convert to this once it's released.

martinemde commented:
I switched the name from `mandate` to `expect`. I used a single commit so we could revert if we decide not to go that route. I personally prefer it.

I'm doing one more pass on the guides to try to find non-obvious uses. Otherwise this is ready for code review (and please call out any documentation that you think should be fixed, if I missed it)

dari-us commented:
@martinemde Yeah, global config for logging and raising/ignoring via method call is fine by me and I think makes sense for the broad majority of apps.
What I mean in particular with "does not play nicely" is this case:

```
action_on_unpermitted_params = :raise
# route PATCH /users/1 => users#update
params.expect(user: [:name, :age]) # raises, b/c params contains unpermitted param :id
```
I think at this point, `:raise` basically kills your app and `:log` produces lots of noise. Also in a nested context like `/groups/1/users`, the `create` action also logs/raises.
To improve the situation, one would probably move the actual filtering that `permit` does into an internal method, say, `apply_filters`, and then expect could use that, so `unpermitted_keys!` is not called automatically.
Also, `unpermitted_keys!` needs to accept an argument to trigger `:raise` by hand and use global config as fallback. 



On another note, sadly the use case how to replace `params.fetch(:x, {}).permit(...)` is not covered, but I guess this is out of scope for this PR. At least I don't know a good way how to incorporate it.

martinemde commented:
@dari-us, ok, I understand now what you mean. The fact of how permit works when applied to params "root" means it would consider any other params to be unpermitted. This explains something I was noticing on rubygems.org, which is that dev was logging more unpermitted params than usual.

Possible fix, but maybe too ugly: `id, nested_id, table_params = params.expect(:id, :nested_id, table: [:attrs])`
I think maybe your proposed solution is the best. Let's consider this a bug with this implementation until we can fix it, since it's not the expected outcome.

As for the other methods, I think you're right. The `fetch.permit` pattern could be `allow(user: [:name, :age])` which would return `{}` on missing key. Then it uses the same pattern as expect but communicates that, just like many mocking frameworks, allows might or might not happen and that's ok. (I prefer allow over accept because accept sounds a lot like expect, just transpose a syllable).

dari-us commented:
@martinemde Actually `expect` and `accept` was intentional as to not confuse it with `permit`, since `permit` and `allow` are kinda samey, but yeah, naming stuff is hard, so I don't know.

I think in order to account for `action_on_unpermitted_parameters`, `expect` (and later on `allow`) needs to iterate each of its keys and call `unpermitted_parameters!` for each sub-hash, f.e.:

```
user_params, group_params = params.expect(user: [:name], group: [:name])
# Here, expect needs to check user_params and group_params each for any unpermitted params and take action

# Maybe something like this
def expect(*filters)
  ...
  values = params.require(keys)
  if self.class.action_on_unpermitted_parameters
    keys.zip(values).each do |k, v|
      next unless v.is_a?(self.class) # value could also be a scalar, if so, ignore
      self[k].unpermitted_parameters!(v)
    end
  end
  values.size == 1 ? values.first : values
end
```

Same will apply to `expect_only` and `allow_only`, but there `unpermitted_parameters!` is unconditionally called with `:raise`

dari-us commented:
@martinemde 
I made changes to Parameters, which passes all the tests (except `keys to arrays: return permitted params in hash key order`, which I disagree with, see my comment). I don't know how to contribute "the right way" or if it is too early for that, but this is what I've come up with:

https://github.com/dari-us/rails/compare/main...dari-us-implement-safe-parameter-filtering#diff-99dab0dfb4d0cfa044997480d5aa1b100dc60347a10cedd6d8f7a0395f6a6efd

Feel free to incorporate / beautify this

martinemde commented:
Thanks for your help @dari-us. I'll incorporate this today.

martinemde committed:
`Parameters#expect` safely filters and requires params

martinemde commented:
Rewritten with your commit as base. I fixed the non-passing test and was able to simplified somewhat. I added expect and allow tests and they are passing.

I replied about your concerns about that one test behavior. I'll copy the code here so we can discuss if anything should be done about it in `expect`. However, I think it should just work the same as permit works right now, which is this (using latest published rails):

```ruby
irb(main):006> params = ActionController::Parameters.new({ user: [{ name: "a" }, { name: "b" }] })
=> #<ActionController::Parameters {"user"=>[{"name"=>"a"}, {"name"=>"b"}]} permitted: false>
irb(main):007> user_params = params.permit(user: [:name]).require(:user)
=> [#<ActionController::Parameters {"name"=>"a"} permitted: true>, #<ActionController::Parameters {"name"=>"b"} permitted: true>]
irb(main):008> Rails.version
=> "7.1.3.4"
```

```ruby
irb(main):015> params = ActionController::Parameters.new({ person: { name: "Martin" }, pie: { name: "Pumpkin" } })
=> #<ActionController::Parameters {"person"=>{"name"=>"Martin"}, "pie"=>{"name"=>"Pumpkin"}} permitted: false>
irb(main):016> params.permit(person: [:name], pie: [:name]).require([:person, :pie])
=> 
[#<ActionController::Parameters {"name"=>"Martin"} permitted: true>,
 #<ActionController::Parameters {"name"=>"Pumpkin"} permitted: true>]
```

dari-us commented:
@martinemde In my mind, returning either `Parameters` or an `Array<Parameters>` should be considered a bug, read on for a demonstration that clarifies this.
Seeing as `expect` aims to replace the `require.permit` pattern, you have to think like this:
When you `require`, you expect an instance of `Parameters`, on which you then `permit` some scalars, i.e. the end result is `Parameters` with `@permitted = true`.
So there is no way you ever expect `require` (aka the root key) to return an `Array` (or anything other than `Parameters` for that matter).
Translating this into `expect`, `expect(user: [:name])` should be equivalent to what devs expect from `require(:user).permit(:name)` according to the "old" Rails docs.

Here is an executable test case demonstrating, that `permit(user: [:name])` is deficient to account for single instances vs Arrays.
As can be seen below, there are 2 variants of `Parameters` for the `User has_one :bio`, one time with an `Array<Hash>`, one time with a singular `Hash`.
Using the ***same*** filter on both of these `Parameters` returns an `Array<Hash>` for the invalid one for `bio_attributes` (when `Hash` is expected and the only valid option), which crashes when assigning it to `User` instance.

Is there a rationale in the docs for returning either `Hash` or `Array<Hash>` for these kinds of nested permits? Seems really error-prone to me.
Your proposed solution of making it explicit that an `Array` is expected via the use of `expect(user: [[:name]])` seems like a really good solution to me. And yes, I absolutely want to be strict about these things. I want to control, whether a key in `Parameters` points to

1. a scalar
2. nested `Parameters`
3. an `Array`

One caveat, `Parameters.new({ attrs: { "0" => { name: "a" }, "1" => { name: "b" } } })` contains a `Hash` on the surface level, but this actually translates into an `Array`. It is also permissible by `permit(attrs: [:name])`, which explicitly allows numbered Hash keys...
I think, such a structure should also ***only*** be permitted by use of `permit(attrs: [[:name]])`.

And another huge caveat: this is a breaking behavioral change, but is actually needed in order to fix the bug demonstrated by below test case.

Maybe only `expect` and `allow` should behave like this.

```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails"
  # If you want to test against edge Rails replace the previous line with this:
  # gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3", "~> 1.4"
end

require "active_record"
require "action_controller"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
    t.string :name
  end

  create_table :bios, force: true do |t|
    t.references :user
    t.text :bio
  end
end

class User < ActiveRecord::Base
  has_one :bio
  accepts_nested_attributes_for :bio
end

class Bio < ActiveRecord::Base
  belongs_to :user
end

class BugTest < Minitest::Test
  def filter_params(params)
    params.permit(user: [:name, { bio_attributes: [:bio] }]).require(:user)
  end

  def setup
    params_hash = { "user" => { "name" => "itsame" } }
    @invalid_params_hash = params_hash.deep_merge("user" => { "bio_attributes" => [{ "bio" => "it is myself" }, { "bio" => "hello, it's just me" }] }).with_indifferent_access
    @invalid_params = ActionController::Parameters.new(@invalid_params_hash)

    @valid_params_hash = params_hash.deep_merge("user" => { "bio_attributes" => { "bio" => "it is myself" } }).with_indifferent_access
    @valid_params = ActionController::Parameters.new(@valid_params_hash)
  end

  def test_same_filter_ensures_data_integrity_for_nested_attributes
    valid_params = filter_params(@valid_params)
    assert_equal @valid_params_hash[:user], valid_params.to_h
    assert User.create(valid_params)

    invalid_params = filter_params(@invalid_params)
    assert_equal @invalid_params_hash[:user], invalid_params.to_h
    # raising NoMethodError instead is really bad!!!
    assert_raises ActionController::ParameterMissing do
      User.create(invalid_params)
    end
  end
end
```

dari-us commented:
@martinemde 
I updated my code to incorporate your suggestion to explicity `expect`, `allow` Array-like structures, see again:

https://github.com/dari-us/rails/compare/main...dari-us-implement-safe-parameter-filtering#diff-99dab0dfb4d0cfa044997480d5aa1b100dc60347a10cedd6d8f7a0395f6a6efd

I introduced `safe_hash_filter`, `safe_actionable_permit`, `array_of_filters?` to account for explicit Array-like structures

dari-us commented:
@martinemde
You could copy/paste the three files I changed in main, they are all passing.
You can then git diff to view my changes compared to your code ^^

dari-us commented:
@martinemde
On a side note, I find it conceptionally strange, that the `*_only` methods are the customizable ones in terms of handling of `unpermitted_parameters`, when their name actually suggests that they are the strict one of each pair. 
What motivates this design choice?

Otherwise, I like the beautifications you've made!

martinemde commented:
I pushed up a version that solves the explicit array requirement and adds tests for it.

As for the `_only` design choice: To me the methods that expect_only or allow_only a certain set of params are the ones that naturally care at all about unpermitted params. In my mind, it doesn't make sense that permit, expect, or allow would yield unpermitted params, because I'm not telling it to be strict about which params are allowed.

I see the `_only` methods as akin to `fetch` where the plain methods are more like `values_at`.

I was also thinking that the explicit array in params should (at some point) become a rails upgrade setting so that it becomes the default.

dari-us commented:
@martinemde Agreed, seeing as the explicit Array filter solves another class of `NoMethodError` caused by param-tampering, it should definitely be the default at some point and the old behavior should become deprecated.
The code looks really good.

Is this ready for final review or is anything still missing?

martinemde commented:
@dari-us I think the docs need to be updated again. I last went through them when we were only adding expect. I also need some coverage for permit_only and all of the _only methods with block arguments. I should be able to get this today.

MatheusRich commented:
@dari-us, this PR has changed quite a bit since I first saw it. Can you ensure the description still matches the contents?

dari-us commented:
@MatheusRich I am not sure how comprehensive the description needs to be (assuming you are referring to the first post), however, @martinemde also made changes to the docs in this PR and as far as I can tell, those are quite comprehensive.

Changes from the initial description:
1. Renamed `mandate` to `expect`
2. Added support for the `fetch(:x, {}).permit(...)` pattern via the method `allow`
3. Added `permit_only`, `expect_only`, `allow_only`, that each work like their counterparts without `_only` with the difference being, that they `raise UnpermittedParameters` if no block is given, or yield the unpermitted keys to the given block.
4. Solved another class of `NoMethodError` by requiring explicit array filters for `expect`, `expect_only`, `allow`, `allow_only`; an in-depth problem description is [here](https://github.com/rails/rails/pull/51674#issuecomment-2273177879). An explicit array filter looks like `params.expect(users: [[:name]])` (i.e. an array of users (or hash with numbered keys) is expected, each with an allowed attribute of `:name`, note the double array brackets)

MatheusRich commented:
@martinemde @dari-us the description is good for now. It was still mentioning `#mandate` which is not part of the implementation anymore. Given the change in the initial scope, I think we should get some feedback from the core team before proceeding. 

cc @dhh, as you've given the green light for the initial API.

martinemde commented:
@dari-us Thank you for the response. I updated the description a little, though I was doing it on my phone at the time so I'll make another pass at it now.

@MatheusRich, @dhh, let me know what you think and if there's anywhere that we should update. I think it's also possible to split the `_only` methods out of this PR.

I think the explicit array behavior is absolutely necessary with this PR because `expect(user: [:name, :age])` would otherwise  open us up to a variation of the original problem I was trying to solve: you could pass an array `[{name: "me", age: 40}]` in place of the expected hash, breaking the type contract that I was trying to fix.

dari-us commented:
@MatheusRich Any updates from Rails core? Is this still on the radar? It's been a few weeks and I'd love to get this going.

dhh commented:
I absolutely recognize the underlying problem here, but I think we've gone a bit off track trying to solve every possible problem permutation, and as a result we've ended up with far too many methods, which leads to much higher conceptual overhead. I don't think that trade is worth it.

I'm willing to add 1 method that captures the require().permit() pattern and deals with the original issue. Everything else should simply raise a parameter exception, which we can capture and discard by default. The annoyance with improper parameter formatting is that it causes noisy exceptions that just don't matter. So the real problem here is how do we get rid of those without harming the development experience,

dhh commented:
Here's a quick sketch:

```ruby
class UsersController < ApplicationController
  # POST /users?name=X will raise ActionController::UnexpectedParameter
  # By default, ApplicationController should have
  # rescue_from ActionController::UnexpectedParameter, with: -> { head :bad_request }
  def create
    User.create! params.expect(user: [ :name, :age ])
  end
end
```

This should not raise on params being `{ user: { name: "x", age: "20", address: "y" } }`. The only thing that'll cause an unexpected exception is if user is missing. Even `{ user: { name: "x" } }` (where age is missing) should not raise. All we're interested in is getting the first key required, and the subsequent keys permitted, without raising a 500 if that's not right.

Together with this, we should add a default public/400-bad-request.html file.

dhh commented:
Only additional method I'd be willing to entertain in addition to #expect is #expect!, which would then raise instead of just swallow the exception. So it would cause a 500 instead of a 400. This might be useful for, say, internal APIs where you really do want to know if a malformed request is being attempted by, say, your own mobile apps.

But we should keep the original mission top of mind otherwise: Get rid of exceptions from malformed requests so that they don't gunk up reporting and alerting pipelines. And we just need to get rid of the majority of it, not necessarily every imaginable case.

dari-us commented:
@dhh
Thanks for the feedback.
Quick sketch why "adding one method" sadly isn't enough:

```
class UsersController < ApplicationController
  # POST /user?user[name]=me&user[bio_attributes][][bio]=evil&user[bio_attributes][][bio]=haxor
  # procudes Parameters { user: { name: "me", bio_attributes: [{ bio: "evil" }, { bio: "haxor" }] })
  def create
    # params.expect succeeds, but assigning it to User.has_one(:bio) instance will raise 500
    # because the filter accepts both an Array of Hashes `bio_attributes` with key `bio` and a singular Hash with key `bio`
    User.create!(params.expect(user: [:name, bio_attributes: [:bio]]))
  end
end
```

Another 500 due to param tampering sadly remains unfixed with "keeping everything as it is just adding one method".

Also, it sounds to me, that you want to discourage/deprecate `action_on_unpermitted_parameters` due to the non-descriptive errors it produces, is that correct? Fine by me, just curious.

dhh commented:
Again, we don't need to solve all permutations of this problem. We just need to solve the most common ones. Chasing a 100% solution is going to lead to a complicated and hard to understand API in service of edge cases. That's the trade-off I'm not willing or interested in making.

 Yeah, only just see `action_on_unpermitted_parameters` now. Not a fan at all. This is not the kind of setting that should be controlled at the top level, and I also think it's an awkward API in any case. Would like to see that go away.

I want parameter exceptions raised by #expect to be caught and 400 returned. If you want an exception, you can use #expect!. It should be evident at the call-site what happens.

dari-us commented:
Not gonna argue whether `accepts_nested_attributes_for` is an edge case or not, at least "flat" parameters can be safely `expect`ed then.
For all other cases, I will use a custom implementation/verification.
What do you think, @martinemde ?

dhh commented:
Yeah, let's keep this on track, because I absolutely would argue that `accepts_nested_attributes_for` was a mistake, and I'd love to remedy that some day as well 😂. But totally cool to extend what we have in the default box here with a SuperStrongParameters gem or the like that can wrap our basic defaults with a much wider API.

martinemde commented:
Thanks for the feedback @dhh.

It sounds like I can refocus on getting rid of annoying 500s that should be 400s.

In service of that focus, let's do the following:
1. drop `allow`, since that's focused on optional parameters, a replacement for `params.fetch(:key, {}).permit`.
2. drop the `_only` methods, since those are addressing `action_on_unpermitted_parameters` which is a separate issue.
3. (?) keep the `[[:array, :of, :hashes]]` since that's now much more likely to cause 500s with `expect` syntax. (@dhh absolutely let me know here if I misread your take on this part)
4. (?) do we add `expect!` to raise when the params are given in a bad format? (I'm not sure it's all that different from ParameterMissing, which we will raise from `expect` and you can rescue if needed. In both cases it's a params that is not as expected (both nil and string are the wrong type which a hash is expected))

If we want to consider each of these separately, I can crunch this down to just `expect` and then open a new PR for `allow` and the `only` methods so we don't have to boil the ocean here.

dhh commented:
@martinemde I think we can take multiple stabs at this. Most important is to get the new #expect syntax in place, and to have that start generating 400s using a new public/400-bad-request.html template that's automatically configured by ActionController::Base using rescue_from.

Remind me of point 3?

For point 4, I'd love to have expect! as a way to transition off `action_on_unpermitted_parameters`. I think that top-level toggle was a mistake, and going forward, it should be clear at the call-site what's going to happen.

martinemde commented:
point 3: `expect(user: [:name, :age])` implicitly allows either `{ user: { name: "Martin" } }` or `{ user: [{name: "Martin"}, {name: "DHH"}] }`. If we introduce `expect(user: [[:name, :age]])` for the second case where an array is expected, then the first case can continue to be written `expect(user: [:name, :age])` and only allow non-arrays to match. The catch here is that `permit` allows the same format and we need to decide if we deprecate the old loose matching.

point 4: it sounds like actually we should replace `expect_only` with `expect!` and keep the same behavior as if `action_on_unpermitted_parameters` is true. That's what the `_only` methods do currently

dhh commented:
Point 3: I see. I like the disambiguation, but we only apply that to this new method. I don't want to break/deprecate anything on the existing methods. Because, really, going forward, our recommendation will almost always be that everyone should be using #expect all the time. So it's not worth the breakage or forcing function on the past API.

Point 4: That sounds great. I'd actually like for #expect/! to entirely ignore `action_on_unpermitted_parameters`. This is a new API, so we can cut a clean break here. For #expect/!, you'll always be able to tell from the call-site what happens. If you bang, there's an exception and a 500. If you don't, there's a 400.

So we're essentially soft deprecating #permit and #require. Don't want to actually deprecate, and force everyone to upgrade. But in docs and whatever, we just focus on #expect going forward.

Are there any common cases that #expect does not handle where someone would still want or need to use #permit and #require?

martinemde commented:
@dhh, I think it covers everything except the cases we were aiming to fix with `allow` or more specific cases where you don't want the keys required.

I think the simplest way to get this merged is to separate the `permit!` behavior into a now PR, which I did here in #52737. This lays all the groundwork for `expect` and `expect!` and covers the use cases where permit would still be useful. It makes it immediately possible to ignore the `action_on_unpermitted_parameters` and use the new API. I see this as a separate issue compared to what expect is trying to solve.

I'm going to build the narrowed `expect` and `expect!` on top of that commit, constraining the `[[:key]]` pattern only to expect for now. I wrote it such that it's easy to switch later if we want to make it a rails version upgrade setting.

dhh commented:
Hmm, maybe we're misunderstanding each other. I don't actually think we should have any function that raises when there are extra parameters. I only want one that raises when the required parameter is missing and another that does 400 when that require parameter is missing. So don't think adding #permit! makes sense.

We'll add #expect, which should cause a 400 when the required part of the expect is missing so `expect(user: [ :name, :age ])` results in the 400. Whereas `expect!(user: [ :name, :age ])` results in a 500 when user is missing (but not if either name or age is missing).

martinemde commented:
@dhh, thank you for clarifying, I think I stumbled over this a little. I closed the PR because it's actually not what we were aiming for. Please ignore the extra noise.

Can you clarify the use case for `expect!`? I imagined it was related to the `action_on_unpermitted_parameters` but it sounds ilke that's unrelated.

From my vantage point, raising `ParameterMissing` allows the user to decide how to behave when the params are malformed for whatever reason. I'm not sure that distinguishing between missing and wrong type is right. Instead, I'd expect people that care about that to rescue `ParameterMissing` and then handle it however they want if they don't want the 400.

dhh commented:
My thinking is this. We're doing all this work to get rid of a bunch of 500s related to malformed requests coming from the world. We generally don't care about folks URL hacking and producing these nonsense requests. We just want to show them a 400. Just like we'll show them a 404, without logging an exception, if they do /users/<id-that-doesnt-exist>.

That's what the new #expect method should largely solve (not withstanding the edge cases on nested assignments, as discussed above). Just throw away malformed requests by serving them a 400 and don't log an exception for it.

Then there's the case where you're not expecting requests from the world. You control the whole chain, and you want an extra layer of protection that you catch any malformed requests, because that means you made a mistake. That to me is the internal API + mobile app case. You control both sides, you don't want malformed requests ignored with a 400. You want a 500, and you want it logged. That's what #expect! is there for.

#expect! should raise a unique exception that's bubbled all the way up, such that it produces a 500, and such that it'll get captured by the likes of Sentry. I don't think we can use ParameterMissing here for that, since we already use that for other cases?

#expect will technically also raise an exception, but we'll have a preconfigured rescue_from that turns that exception into a 400, and doesn't bubble.

Does that make sense?

martinemde commented:
@dhh, got it. Private/internal API use case explains it perfectly. I can do that.

Should `expect!` also be strict about unpermitted params (extras that shouldn't be there). The new `expect!` is intended to be rather strict about what it receives. In that case I would propose we raise the existing error `ActionController::UnpermittedParameters` when something is there that wasn't permitted, and `ActionController::MalformedParameters` when something is the wrong type. `expect` would just consider malformed to be the same as missing, but `expect!` would disambiguate.

dhh commented:
No, I generally don't find strict useful. Who cares what extra stuff is being submitted, as long as you get what you need to process the request. I think we currently have a logging function in development that logs this, which I sorta can tolerate, but even that I often find annoying. So let's just ensure that expect! only raises if you don't get what you asked for (which is the same as what required used to do).

What is required() currently raising? Does feel like it's that which we want.

martinemde commented:
> What is required() currently raising? Does feel like it's that which we want.

`ActionController::ParameterMissing` which is the same thing `expect` would raise if the params are malformed, and already what is handled by rails as a 400.

dhh commented:
Ah, perfect. Then we're good with that. Then it's just #expect! that needs another exception such that it'll produce a 500 instead of 400. Hmm. Maybe `ActionController::ExpectedParameterMissing`?

dhh commented:
Alternatively we could keep the same exception but add a flag on it whether to swallow (400) or bubble (500)? Although having separate exceptions is probably easier.

dari-us commented:
@dhh Your stance on nested parameter assignment is conflicting with what you said in regards to @martinemde Point 3, which is basically the same issue, i.e. clear disambiguation between expecting `Parameters` (`params.expect(user: [:name])`) versus `Array<Parameters>` (`params.expect(user: [[:name]])`), as `params.expect(user: [:name])` with the current implementation of `permit` allows both `{ user: { name: "Martin" } }` and `{ user: [{ name: "Martin" }, { name: "DHH" }] }`, which you said you want to disambiguate.

Solving one but not the other is rather inconsistent IMHO.

martinemde commented:
@dari-us I think we _are_ including the explicit array format for `expect` and `expect!`.

martinemde commented:
@dhh, I'll implement this as described. Thank you for clarifying.

I'll come up with a name and we can change it in review easily.

In summary:
* `expect`: strict about types, explicit array format, not strict about excess params that are ignored, raises ParameterMissing which gets picked up and rendered as a 400. Add a 400 html template.
* `expect!`: same as expect, but if a parameter is the wrong type, not just missing, it will raise a new error, somewhat preserving the behavior of permit right now, but raising a better error than `ArgumentError` so it helps fix internal APIs, etc.

dhh commented:
Yes, I think that all sounds correct. Thanks for working on this!

dari-us commented:
@martinemde 
I made a fresh fork of rails and applied the changes you made and implemented the methods as requested.
So this should almost be the final thing:
https://github.com/rails/rails/compare/main...dari-us:rails:params-expect

In summary:
- Removed `expect_only`, `allow`, `allow_only`, `permit_only` and their corresponding tests
- Created `class ExpectedParameterMissing < ParameterMissing; end`
- `expect!` is exactly like `expect`, but raises `ExpectedParameterMissing` instead of `ParameterMissing`
- Reverted changes to raising/logging unpermitted parameter methods
- Some documentation updates (hopefully didn't miss anything)
- Tests for `expect!` are missing, but the only thing that changes from `expect` is the raised exception

dari-us committed:
Add public/400.html file rendered on bad request

dari-us committed:
`Parameters#expect!` raises unhandled exception on missing params

martinemde commented:
@dari-us, hmm, you interpreted the `expect!` behavior slightly differently than I did. I saw a distinction between missing and wrong type being useful. Maybe I'm overthinking it. I had started down a path of defining a different params processor that detects when we get the wrong type for an expected filter and raised `"`[]` filter expected Array, got Hash"`.

Yours is certainly much simpler, but it blurs missing and malformed. It is very similar to adding `rescue ActionController::ParameterMissing` in your action anywhere you would use `expect!` which doesn't provide any extra information about whether your internal API is being used incorrectly.

martinemde commented:
If you're amenable, this PR now contains only the `expect` method. We could merge this and then continue with expect! once we have a finalized behavior for `expect`. I think there's still a little confusion.

If all we need is @dari-us' approach, then it's simple to add and @dari-us can simply make an additional PR. If we want to catch and report expected types separately than missing/blank params and not just raise for all of them, then we need a more thorough re-plumbing of the parser. It currently has a lot of places where incorrect types are ignored (it effectively cherry-picks the type it's looking for in almost all cases, ignoring anything else that's incorrect.)

dari-us commented:
> if a parameter is the wrong type, not just missing, it will raise a new error, somewhat preserving the behavior of permit right now

"Preserving the behavior of permit right now" to me sounded like discarding wrong types without being specific about what went wrong.
So `{ user: { name: "Me" } }.expect(:user)` would produce a message saying, that `user` was missing or empty, i.e. not part of the hash. You want something like `user needs to be scalar, but was Hash`, right? Not sure if this could be simplified to saying `param is missing or the value is empty or unexpected: user`.

Saves the trouble and complexity of rewriting all filter methods and should hopefully be enough for a dev to investigate the issue. I thought the main benefit of `expect!` is that it is not rescued and turned into a 400, but instead produces a 500.

Of course you can `rescue` yourself, but then again, you could technically also `rescue` all your controller actions instead of using `rescue_from`, but it's more convenient this way.

(Random nitpick: You used "premit", i.e. swapping "r" and "e" somewhere in your documentation)

@dhh Are we on the wrong track here regarding `expect!`?

martinemde commented:
@dari-us typo fixed, thanks.

Let's come up with some examples where I think `expect!` would trigger an error. I'll use the same shorthand. I think we all understand what is meant.

```ruby
{ user: { name: "martin" } }.expect(:user) # UnexpectedParameter: `:user` expected scalar, got Hash
{ user: { name: "martin" } }.expect(user: [[:name]]) # UnexpectedParameter: `[[:name]]` expected Array, got Hash
{ user: ["martin"] }.expect(user: [:name]) # UnexpectedParameter: `[:name]` expected Hash, got Array
{ user: ["martin"] }.expect(user: [[:name]]) # UnexpectedParameter: `[:name]` expected Hash, got String (?I don’t know?)
{ user: "string" }.expect(user: [:name]) # UnexpectedParameter: `[:name]` expected Hash, got String
{ user: { name: { first: "martin" } } }.expect(user: [:name]) # UnexpectedParameter: `:name` expected Scalar, got Hash
{ user: { name: { first: "martin" } } }.expect(user: { name: [] }) # UnexpectedParameter: `:name` expected Scalar, got Hash
```

So do we go deeply into the params to find unexpected formats, do we ignore them (sometimes they will be ignored regardless because

dhh commented:
I was also just thinking that expect! would catch and reraise. I hadn't considered distinguishing between missing vs wrong type. I'm skeptical that's necessary.

dhh commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,50 @@
+*   Introduce safer and more explicit params handling methods.
+    - `params.expect` replaces `params.require(:table).permit(:attr)`
```
Would expand to `params.expect(table: [ :attr ])` for this headline example.

dhh commented:
I think the API looks correctly now. Maybe @rafaelfranca could have a 2nd look.

MatheusRich commented on actionmailbox/app/controllers/rails/conductor/action_mailbox/inbound_emails_controller.rb:
```diff
@@ -31,7 +31,7 @@ def new_mail
 
       def mail_params
-        params.require(:mail).permit(:from, :to, :cc, :bcc, :x_original_to, :in_reply_to, :subject, :body, attachments: [])
+        params.expect(mail: [:from, :to, :cc, :bcc, :x_original_to, :in_reply_to, :subject, :body, { attachments: [] }])
```
Do we need to explicitly add the hash brackets here?

martinemde replied:
Yes, not valid ruby syntax otherwise.

martinemde replied:
Oh wait, wow, you're right. I didn't realize that _was_ valid.

martinemde replied:
ok, I will undo the changes where I added the hash brackets. Somehow I was so sure this was wrong that I never tried it.

MatheusRich replied:
Is it? Maybe I'm misunderstanding it, but this works:

```rb
def args(...) = puts(...)

args(a: [:b, c: []])
# {:a=>[:b, {:c=>[]}]}
# => nil
```

martinemde replied:
Fixed and pushed! I'm really happy you called this out. I thought this was ugly but necessary. Now it looks way better and I feel much more happy with the suggested syntax.

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
`ExpectedParameterMissing` can simply inherit from `ParameterMissing`, as inheritance does not affect `rescur_responses.

In `def expect!` namespacing the exception classes is unnecessary.

In `def permit_filters`, unpermitted params handling should be reverted to `unpermitted_parameters! if check_unpermitted && self.class.action_on_unpermitted_parameters`

`def hash_filter` needs a kwarg named `check_unpermitted` that is passed back to `permit_filters` so nested params are **never** logged/raised for `expect`.

Also, in many places, methods are called with kwargs with the same local variable name, like `func(stuff: stuff)`. I thought the preferred way was to always `func(stuff:)` in those cases.

martinemde replied:
1. It seems a little odd since it is the opposite behavior of `rescue`. For exaple, `rescue StandardError` catches `RuntimeError` so I wouldn't use `RuntimeError` as a "non-handled" error when `StandardError` is always rescued. It seems to violate expectations.
2. 👍 
3. 👍 
4. Previous behavior was to check unpermitted at every level of params. With `expect`, we decided to exclude the root level out of necessity. Did I miss that where we decided to ignore unpermitted params behavior for `expect`? I just thought it was something people could choose or not.
5. The inconsistent splatting is necessary and intentional. It's the reason that I created a new method instead of adding kwargs to permit or expect. Once you declare a kwarg in the method definition, any other kwargs are ignored. If you add `**kwargs` to the method definition, you alter the order of the arguments which breaks the contract for order of returned values. You can't achieve what we want while still splatting the args to `permit_filters`.
6. That's not supported on ruby 2.6 or 2.7. Does Rails still aim to support 2.6/2.7? In a cursory glance I don't see this format being used so I was playing it safe.

dari-us replied:
1. Yeah, but that's the way Rails handles it 🤷‍♂️ I guess it's so they can `rescue_responses[exception.class.name]`. Otherwise you need to (safe) constantize each key of `rescue_responses` and check its ancestors.
4. Feels kind of weird to have a variable named `check_unpermitted` (sounds boolean) with values `:log` or `:raise`. Maybe rename to `on_unpermitted`? That way, it's sensible to have values `nil`, `:log` and `:raise`.
5. Thanks for the explanation I wasn't fully aware of that.
6. Not sure if this will be backported to the Rails 7 releases or only available from 8.0 onwards. If the latter, I think only Ruby 3.1 is supported, so the "new" syntax is safe in this case. But yeah, better play it safe, then.

simi replied:
6. New features are not usually backported, and even it would be, [Rails 7 requires Ruby 3.1+](https://github.com/rails/rails/blob/a11f0a63673d274c59c69c2688c63ba303b86193/rails.gemspec#L12).

martinemde replied:
@simi thanks for finding that. Updated (and it looks much better).

dari-us reviewed (commented):
Found some things that may be worth changing

martinemde commented on actionpack/test/controller/parameters/log_on_unpermitted_params_test.rb:
Is this the intention? I didn't think we were changing this behavior at all except to avoid root level warnings . Using `expect` on the root of the params was much likely to trigger unnecessary warnings in dev. If we're disabling it entirely, is the intention that people who want that behavior fall back to permit/require?

martinemde commented:
@dari-us Thank you for thoroughly looking through the commit. I fixed most of the things you mentioned.

I reread @dhh's comments after responding above. `expect!` was mentioned as a way to transition off of `action_on_unpermitted_parameters`.

Correct me if I'm wrong, but therefore `expect!` will always raise on unpermitted params and `expect` will never raise on them. Correct?

martinemde commented on actionpack/test/controller/parameters/raise_on_unpermitted_params_test.rb:
```diff
@@ -24,10 +24,82 @@ def teardown
+      author_id: 1,
+      id: 1,
+      book: { pages: 65, title: "Green Cats and where to find them." })
+
+    # this one clearly should not raise, everything is covered.
+    assert_nothing_raised do
+      params.expect!(:id, :author_id, book: [:pages, :title])
+    end
+
+    # This one is the tricky one. Should it raise?
+    assert_nothing_raised do
+      params.expect!(book: [:pages, :title])
+    end
+
+    # this one clearly should raise, both a root key and a hash are declared.
+    assert_raises(ActionController::UnpermittedParameters) do
+      params.expect!(:id, book: [:pages, :title])
+    end
+
+    # This one is not quite as obvious. Both scalar root keys are declared
+    assert_raises(ActionController::UnpermittedParameters) do
+      params.expect!(:id, :author_id)
+    end
+
+    assert_raises(ActionController::UnpermittedParameters) do
+      params.expect!(:id)
+    end
+
+    assert_raises(ActionController::UnpermittedParameters) do
+      params[:book].expect!(:pages)
+    end
+
+    assert_nothing_raised do
+      params[:book].expect!(:pages, :title)
+    end
+
+    assert_nothing_raised do
+      params(book: {}).expect!(:pages, :title)
+    end
+  end
```
Ok, this test fleshes out the potential behaviors for `expect!`.

I'm guessing that I'm overthinking this again, but this gets quite complex.

The simplest answers:
1. You must specify all keys at all levels.
   This code fails: `User.find(params.expect!(:id)).update(params.expect!(user: [:name]))`
2. Unpermitted root keys are always ignored.
   `params.expect!(:id)` always works, but `params.expect(user: [:name])` raises on `user[age]=` being present but `params[:user].expect(:name)` would not raise.

martinemde replied:
I'm removing all unpermitted parameter checking from expect/!

dhh commented:
I was thinking expect! would only raise if the request is malformed or the root element is missing. Not if there are additional parameters that are not being used.

martinemde commented:
@dhh, ok. Do both expect and expect! ignore the `action_on_unpermitted_parameters` config? It would just be reserved for `permit` then, which seems fine to me.

martinemde commented:
I re-squashed this into 3 reasonably single feature commits. I'm ignoring `action_on_unpermitted_parameters` for `expect` and `expect!` which is the cleanest solution. `permit` remains available if that behavior is desired (and we can follow up later if we need with a `permit_only` on `expect_only` if people want this feature badly enough.)

Please let me know if there's anything else that needs addressing. I think it's very close, if not actually ready.

dari-us commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,50 @@
+    # raises NoMethodError, causing a 500 and potential error reporting
+
+    # After
+    params.expect(person: [:name, :age, pet: [[:name]])
```
The key should be named `pets`, not `pet`

dari-us commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,50 @@
+    We suggest replacing `params.require(:person).permit(:name, :age)`
+    with the direct replacement `params.expect(person: [:name, :age])`
+    to prevent external users from manipulating params to trigger 500
+    errors. A propery 400 error will be returned instead.
```
Change `propery` to `proper`

dari-us commented:
@martinemde
A few more nitpicks, as of the above two review comments, also see my reasoning in regards to renaming kwarg `check_unpermitted` to `on_unpermitted` across all methods. Instead of calling with `false` from `expect`, using `on_unpermitted: nil` would be sensible then, as in "on unpermitted parameters do nothing". That is in line with the alternate values `:log` and `:raise`

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -39,6 +41,15 @@ def corrections # :nodoc:
+  #     params.expect!(:a)
+  #     # => ActionController::ExpectedParameterMissing: param is missing or the value is empty: a
+  class ExpectedParameterMissing < ParameterMissing
+  end
```
When defining a class without body, I think it is preferable to use one-line `class Abc; end`

martinemde replied:
I was just keeping with the style on line 85.

dari-us replied:
I see, no strong preferences here, consistency is the main thing

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -1174,45 +1360,63 @@ def permitted_scalar_filter(params, permitted_key)
       EMPTY_ARRAY = [] # :nodoc:
       EMPTY_HASH  = {} # :nodoc:
-      def hash_filter(params, filter)
+      def hash_filter(params, filter, on_unpermitted: false, explicit_arrays: false)
```
Default values should match default values of `def permit_filters`, i.e. `def hash_filter(params, filter, on_unpermitted: nil, explicit_arrays: true)`

martinemde replied:
I partly disagree. `hash_filter` existed before this, so its the method signature should be defined in a way that it keeps its original behavior. `permit_filters` is new and defines its defaults as the preferred usage in keeping with `expect`.

I partly agree though because I think this means the default should actually be `on_unpermitted: self.class.action_on_unpermitted_parameters`.

dari-us replied:
You have a point here👍

dari-us commented:
To make sure we're all 100% on the same page here:
`expect/!` **only** raises on missing or wrongly typed root keys.
1. `{ user: "haxor" }.expect(user: [:name])` raises
2. { user: { name: "me" } }.expect(user: [:name, :age])` succeeds
3. `{ user: { age: "100" } }`.expect(user: [:name])` succeeds (with empty hash)
4. `{ user: { name: "me" } }.expect(user: [:name, pets: [[:name]]])` succeeds

On a side note, I am not happy about the fact the exception message says "missing or empty" instead of "missing, empty or invalid", since technically it can also raise for non-empty values (example 1). However, since it is internally raised by `require` we cannot tell whether the key was originally given. Don't know if there is an elegant way to raise the same exception class via `require` vs `expect` while keeping the original message of `require` intact.

martinemde commented:
Parameter missing can happen on type mismatches even before this PR, without `expect`. It just happens less often. I pushed a change to the error message to include "invalid".

Example:
```ruby
ActionController::Parameters.new(comments: { text: "Hello" }).permit(comments: []).require(:comments)
```

dari-us commented:
@martinemde Good example, I think changing the message is the best way forward. Seeing as `fetch(:key)` can **only** raise on missing key, but not on empty value, the exception message for that is already "too much", as "value is empty" can be ruled out by using `fetch`, and yet it is still mentioned.
So I see no problem in adding another "case" to the message.

dari-us commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,50 @@
+    # raises NoMethodError, causing a 500 and potential error reporting
+
+    # After
+    params.expect(person: [:name, :age, pets: [[:name]])
```
A third closing array bracket is missing, i.e. `... pets: [[:name]]])`

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -502,23 +514,25 @@ def permit!
-    #       params.require(:person).permit(:name).tap do |person_params|
-    #         person_params.require(:name) # SAFER
-    #       end
+    #       params.expect(:person: :name).expect(:name)
```
Invalid syntax, `(:person: ...)` does not work, same error two lines below.

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -646,20 +683,114 @@ def require(key)
 
-      params.permit!
+    # `expect` is the preferred way to require and permit parameters.
+    # It safer than the previous recommendation to call `permit` and `require`
```
"It" -> "It is"

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -646,20 +683,114 @@ def require(key)
+    # It safer than the previous recommendation to call `permit` and `require`
+    # in sequence, which could allow user triggered 500 errors.
+    #
+    # `expect` is more strict with types to avoid a number potential pitfalls
```
"a number **of** potential pitfalls"

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -646,20 +683,114 @@ def require(key)
+    #     user_params[:pets] # => []
+    #
+    # As the examples show, `expect` requires the `:user` key, and any root keys
+    # similarl to the `.require.permit` pattern. If multiple root keys are
```
"similarl" -> "similar"

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -646,20 +683,114 @@ def require(key)
+    #    params = ActionController::Parameters.new(name: "Martin", pies: [{ type: "dessert", flavor: "pumpkin"}])
+    #    name, pies = params.expect(:name, pies: [:type, :flavor])
+    #    name # => "Martin"
+    #    pies # => #<ActionController::Parameters {"type"=>"dessert", "flavor"=>"pumpkin"} permitted: true>
```
`pies` should be an Array and also should use the explicit array syntax in the corresponding `.expect` call

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -646,20 +683,114 @@ def require(key)
+    #    subject # => #<ActionController::Parameters {"name"=>"Martin"} permitted: true>
+    #    object  # => #<ActionController::Parameters {"pie"=>"pumpkin"} permitted: true>
+    #
+    # Besidse being more strict about array vs hash params, `expect` uses permit
```
"Besidse" -> "Besides"

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -1088,23 +1239,45 @@ def specify_numeric_keys?(filter)
+      #     { pies: [{ flavor: "rhubarb" }, { flavor: "apple" }] }
+      #     { pies: { "0" => { flavor: "key lime" }, "1" =>  { flavor: "mince" } } }
+      #
+      # Previously, arrays were specified the same way as hashes:
```
Maybe change to "When using `permit`, arrays are specified...". Also use `permit` in the below example, since `expect` never allowed non-explicit arrays in the first place.

dari-us commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -646,20 +683,114 @@ def require(key)
+    #     # with nested parameters
+    #     params = ActionController::Parameters.new(user: { name: "Martin", pets: { name: "hack" } })
+    #     user_params = params.expect(user: [:name, pets: [[:name]] ])
+    #     user_params[:pets] # => []
```
The actual and correct behavior is
`uers_params[:pets] # => nil`

martinemde commented:
@dari-us Updated. Thanks again for reading through everything thoroughly.

dhh commented:
Let me know when this is ready for final review.

martinemde commented:
@dhh, I think it's ready. I just pushed all the fixes that @dari-us noticed.

dari-us commented:
Agreed, as far as I am concerned, this is ready for final review.

MatheusRich commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -502,23 +514,21 @@ def permit!
-    #       params.require(:person).permit(:name).tap do |person_params|
-    #         person_params.require(:name) # SAFER
-    #       end
+    #       # params.expect(:person: :name).require(:name)
```
Syntax error here
```suggestion
    #       # params.expect(person: :name).require(:name)
```

martinemde replied:
Oh huh, I thought I fixed that one.

MatheusRich commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -646,20 +679,114 @@ def require(key)
+    # explicitly. Use double array brackets, an array inside an array, to
+    # declare that an array of parameters is expected.
+    #
+    #     params = ActionController::Parameters.new(comment: [{ text: "hello" }, { text: "world" }])
+    #     params.expect(comments: [[:text]])
```
```suggestion
    #     params = ActionController::Parameters.new(comments: [{ text: "hello" }, { text: "world" }])
    #     params.expect(comments: [[:text]])
```

MatheusRich commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -646,20 +679,114 @@ def require(key)
+    #     params.expect(comment: [:text])
+    #     # => ActionController::ParameterMissing: param is missing or the value is empty or invalid: comment
+    #
+    # In order to permit an array of parameters, the array must be defined
```
I don't love the double brackets, but it is what it is. We could, theoretically, check if the key is pluralized and assume the values will be an array, but that probably would cause more harm than good.

martinemde replied:
An alternative that is maybe uglier is utility method like `expect(comments: params_array(:text, :id))`. I'm not sure it's any better, but it solves the problem. `[[:text, :id]]` is the most concise way to do it that I could come up with (if only ruby had a native tuple type).

martinemde replied:
Alternatively, it can be written as: `expect(comments: [ %i[text, id] ])` vs `expect(user: %i[name, ago])`. I avoided this format since it's requires slightly more knowledge of ruby, and is therefore slightly less approachable.

MatheusRich commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -646,20 +679,114 @@ def require(key)
+    #     # => [#<ActionController::Parameters { "text" => "hello" } permitted: true>,
+    #     #     #<ActionController::Parameters { "text" => "world" } permitted: true>]
+    #
+    #     params = ActionController::Parameters.new(comment: { text: "hello" })
```
```suggestion
    #     params = ActionController::Parameters.new(comments: { text: "hello" })
```

martinemde replied:
Thanks for catching these. I've read over all this so many times that its hard to catch the little details.

martinemde commented:
Oops, I screwed up a rebased and lost the method definition. One moment while I fixe it.

martinemde commented:
Ok, sorry about that. Should be ready again. I'm not sure even what I did there 😬

dhh committed:
Merge branch 'main' into params-mandate

dhh commented:
Excellent work, guys! This is a very nice upgrade.

We'll need to also upgrade the templates in the jbuilder gem, since they're used by default. Probably have to have a Rails version toggle where we use expect when 8.0+ and whatever is there before that. @martinemde want to take a stab at that too?

bogdan commented:
Nice change, silents some noise in bug notifications. Should there be a rubocop rule for it?

dhh commented:
> Nice change, silents some noise in bug notifications. Should there be a rubocop rule for it?

No. Don't need to nag people about this. They can upgrade to the new syntax as they see fit and new apps will automatically guide folks in that direction.

dari-us commented:
@p8 I hardly see that as a problem, at least I cannot remember ever seeing `except` used with `params`.
`allow` sounds less strict, i.e. I wouldn't expect it to raise for missing root keys.
I would love to have `allow` as a replacement for `params.fetch(:x, {}).permit(...)` sometime, as this pattern appears in the docs and is not fixed by the `expect` API.

p8 commented:
I can see non-native English speakers or people with dyslexia having problems with this - copying the guide documentation and wondering why things aren't working as expected.

As this API seems to be more strict than `require` and `permit`, maybe a verb that has a more strict meaning would be clearer?

For example `constrain` or `contraints`, which has overlap with [routing contraints](ttps://edgeapi.rubyonrails.org/classes/ActionDispatch/Routing/Mapper/Scoping.html#method-i-constraints).

dhh commented:
Appreciate the feedback here, but this bikesheed has already been painted, @p8. We're moving on to other bikesheeds now 😄

jeromedalbert commented:
I took a stab at updating the jbuilder templates here: https://github.com/rails/jbuilder/pull/573.
